feat(fields): implement Param and CallableParam for handling unmapped parameters that can be referenced during build

leverage-analytics opened the pull request:
This completes the development and testing related to #604.

- Implement new Param and CallableParam field types
- Update handling of **kwargs for base factory `process_kwargs` and `process_kwargs_coverage`. A new helper method was used to determine which fields in a factory are Unmapped types (Params or CallableParam)
- Implement tests for new fields as well as changes to factory build
- Update documentation to include overview of feature functionality with examples

adhtruong commented on polyfactory/fields.py:
```diff
@@ -114,3 +115,121 @@ def to_value(self) -> Any:
         raise ParameterException(msg)
+
+
+class NotPassed:
```
Can the existing `Null` type be used for this?

leverage-analytics replied:
I had originally thought the same thing but, in the end, decided to define a different constant just so that I wouldn't be using the existing Null in an unintended manner. 

I'm happy to make this change.

adhtruong replied:
This is currently used in multiple places to represent this so this fine to reuse

adhtruong commented on polyfactory/fields.py:
```diff
@@ -114,3 +115,121 @@ def to_value(self) -> Any:
+        return cast(T, self.param)
+
+
+class CallableParam(Generic[T], BaseParam[T, Callable[..., T]]):
```
What is the use case this? Param type itself allows new functionality but this seems like adding bespoke syntax for a less common case

I think  supporting `default_factory` or similar on `Param` itself cover this use case without having to deal with arg or kwargs and can unify this into one class

adhtruong replied:
Got you. The use case for a callable makes sense so can use existing functions/faker instance.

I think just allowing a callable that gets resolved should cover this and also makes sense if want to support decorator form like with postgenerated or similar. I think the interface of have a callable passed in build that gets called with params is less clear and is less like other existing fields. This can be done on user side so think is fine to not support 

> That being said, I do really like your proposal for keeping it all in one class and using a default_factory param for the constructor. I can refactor and resubmit a pull request if you like.

I'd keep it in the same PR and update in place. Happy to go with either
1. A separate class that takes in a callable resolved at runtime with args/kwargs passed to function with build time behaviour being fully overridden. This matches behaviour of other fields
2. Combine these and use `default_factory` or similar

adhtruong commented on tests/test_factory_fields.py:
Is there a convention being used for tests using `__` between methods for these new? This convention is not similar to existing tests so not sure should introduce here

leverage-analytics replied:
You're correct; I can fix that.

adhtruong commented on polyfactory/factories/base.py:
```diff
@@ -1016,7 +1048,7 @@ def process_kwargs(cls, **kwargs: Any) -> dict[str, Any]:
             result[field_name] = post_generator.to_value(field_name, result)
 
-        return result
+        return {key: value for key, value in result.items() if key not in params}
```
Does this logic work for when `Param` is used against a factory field? This would raise an error for kwargs required by the model in potential unclear way. Should an error be explicitly raised in this case

leverage-analytics replied:
Sorry, I wanted to make sure I'm understanding what you mean by param is used against a factory field. Could you please provide an example?

Thanks for reviewing my pull request!

adhtruong replied:
```
from dataclasses import dataclass

from polyfactory.factories import DataclassFactory
from polyfactory.fields import Param


@dataclass
class MyModel:
    values: int


class MyModelFactory(DataclassFactory[MyModel]):
    values = Param(1)


print(MyModelFactory.build())
```
take this example. This will fail as `values` will be omitted from final kwargs. Should Param be checked against factory fields? Maybe in `__init_subclass__`?

> Thanks for reviewing my pull request!

And thank you for submitting! Great feature and work on this :)

leverage-analytics commented:
@adhtruong, I have an update ready to update this pull request, but after merging with main, I am encountering issues realted to some typing related to SQLAlchemy factories. Please let me know how you want me to proceed (wait for patch, submit anyways, or something else).

Here's what I'm seeing:

```
check python ast.........................................................Passed
check for case conflicts.................................................Passed
check for merge conflicts................................................Passed
check toml...............................................................Passed
debug statements (python)................................................Passed
fix end of files.........................................................Passed
mixed line ending........................................................Passed
trim trailing whitespace.................................................Passed
ruff.....................................................................Passed
ruff-format..............................................................Passed
codespell................................................................Passed
prettier.................................................................Passed
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

tests/sqlalchemy_factory/test_sqlalchemy_factory_v2.py:95: error: Call to untyped function "JSON" in typed context  [no-untyped-call]
tests/sqlalchemy_factory/test_sqlalchemy_factory_v2.py:98: error: Call to untyped function "JSONB" in typed context  [no-untyped-call]
Found 2 errors in 1 file (checked 139 source files)

pyright..................................................................Passed
Sphinx Lint..............................................................Passed
```

leverage-analytics commented:
@adhtruong Interestingly, when I ran pre-commit separately, it raised the mypy errors, but when it ran as part of the git hook, it passed. Not sure what is happening there.

Anyways, here is the updated feature we have been discussing. Thank you for your patience as I reworked it. From our conversation, I did a little thinking, and I decided to alter the API just slightly from what we discussed. Instead of including a `default_factory` parameter for the constructor of `Param`, I added a bool flag `is_callable`. The reason I decided to this is to provide for more flexibility in the use of the `Param` type. 

I needed a way to understand if the user's intention is to call the passed value (for example passing a lambda that they might not want to be called) but also override that value at build time and still have the same action (call it or not). I didn't want the user to have to pass a temporary placeholder value for the param at factory definition time (just to minimize the amount of code needed to use the feature) so we could understand their intention related to calling or not calling the passed value. So I figured asking with the boolean option would be a simpler interface for the user.

Thanks again for your review and feedback!

leverage-analytics commented:
I'm sorry if I didn't explain myself accurately, but the kwargs that get passed to the callable can only be passed during the declaration of the factory, not at build time. I thought there was too much potential for collision between kwarg names and factory field names to attempt to handle passing kwargs at build time.

Essentially, there are two routes:

- declare a param at factory declaration with `is_callable` set to `True`: This will require either a callable be passed at factory declaration or at build time (or both, giving priority to whatever is passed at build time). In this case any kwargs passed to the `Param` constructor in the factory class get passed to the callable at build time.
- declare a param at factory declaration with `is_callable` set to `False`. This will require any value to be passed either at factory declaration or at build time (or both, giving priority to whatever is passed at build time). In this case, any kwargs passed in the `Param` constructor are invalid, and an exception is raised.

I thought it made sense, from the viewpoint of API consistency, that if we allow overriding `Param` with a non-callable at build time, we should also allow overriding a callable `Param` at build time, too.

leverage-analytics commented:
@adhtruong @guacs let me know if I can clarify or respond on this any further or how you want to proceed. Thanks!

adhtruong commented:
Hey @leverage-analytics , sorry the late reply on this!

I think the callable use case may be a little confusing as doesn't link the params of kwargs closely to the calllable passed in a build time.  I think just having the whole value overriding is fine and keeps overall design simpler. This matches the existing interface of other fields closer e.g. if `Use` is defined on the factory when a value is passed in the args and kwargs are ignored.

If there is a strong use case for  this kind of overriding, then have the original form of separate field types for callable and not makes this difference more clear so usage if consistent at field level.

leverage-analytics commented:
@adhtruong thanks for the feedback! As I see it, our options are:

1) drop support for passing `kwargs` to callable `Param`s (so the user doesn't confuse the two uses of `kwargs`). I think this makes the most sense. I personally don't have a use for `kwargs` to be passed to the callable; I was just trying to accommodate use cases I hadn't thought of.
2) revert to the original pull request with separate Param and CallableParam types. 
3) drop the functionality related to the calling a callable parameter at build time. I'm not a fan of this, as this would mean, at least for my use case, a lot more repeated code in some cases. This would mean a user who wants to call a param at build time would have to pass its value as a build kwarg every time. 

Like I said, I'd prefer the first, but I'm open to supporting the project how the maintainers see best. 

Thanks!

adhtruong commented:
1 sounds good! Thanks for the write up and explanation :)

adhtruong commented:
Just to clarify 1. Would this would looks like

```py
class Param:
     def __init__(self, default: T, default_factory: Callable[[], T)-> None: ...

class Factory(DataclassFactory[MyType]):
      a = Param(default_factory=lambda: 1)

assert Factory.build(a=2).a == 2. # would this error?
assert Factory.build(a=lambda: 2) == lambda: 2 # would this get evaluated?
```

For 3 use case, can't you pass this in evaluated form? Or overridden via `Factory.create_factory(a=Param(default_factory=lambda : 2)`? Or may be misunderstanding this still

leverage-analytics commented:
@adhtruong answered based on my proposal

```python

class Factory(DataclassFactory[MyType]):
      a = Param(default_factory=lambda: 1)

# This example results in an attribute error because no parameter values get mapped 
# into the objects produced by the factory (assuming MyType doesn't have attribute 'a').
assert Factory.build(a=2).a == 2. 

# In the following case, the expression lambda: 2 would be evaluated during build 
# because the user specific a default_factory for the`Param`. 
assert Factory.build(a=lambda: 2)
```

adhtruong commented:
Thanks again for clarifying. Appreciate the patience on back and forth and explaining to help me understand!

I think 3 makes is my preference as most in line with existing fields and result can be achieved by user with 
1. Call a callable before passing in 
2. Or override if need this repeating as a factory which can avoid repetition.

I think can make a minor tweak (commented in line) to make this work for coverage form too

adhtruong commented on polyfactory/factories/base.py:
```diff
@@ -1085,6 +1132,9 @@ def process_kwargs_coverage(cls, **kwargs: Any) -> abc.Iterable[dict[str, Any]]:
         result, generate_post, _build_context = cls._get_initial_variables(kwargs)
 
+        params = cls.get_factory_params()
+        result.update(cls._handle_factory_params(params, **kwargs))
```
Should this having a separate coverage form to handle the callable type? Otherwise it will have the same for all

leverage-analytics replied:
If you take a look at this section, it calls the `to_value` method of the `Param class to determine the handling.

https://github.com/litestar-org/polyfactory/blob/19fe618562d6b87b3e4032a298993f40496f14c3/polyfactory/factories/base.py#L1033-L1048,

leverage-analytics replied:
And then this is how the `Param` method determines what is going on.

https://github.com/litestar-org/polyfactory/blob/19fe618562d6b87b3e4032a298993f40496f14c3/polyfactory/fields.py#L153-L200

adhtruong replied:
So question here is if should be static for the whole of `process_kwargs_coverage` or not. Currently if have param with `lambda: random.choice([1, 2])` and field `Literal['a', 'b', 'c']` then output would be

```python
{"foo": 1, "bar": 1}
{"foo": 1, "bar": 2}
{"foo": 1, "bar": 3}
```
as param is only evaluated once. If evaluated each run then would give something like

```python
{"foo": 2, "bar": 1}
{"foo": 1, "bar": 2}
{"foo": 2, "bar": 3}
```
Either or is fine I think

leverage-analytics replied:
Yes, I am indifferent here

leverage-analytics commented:
So, would you like me to move forward with your preference of choice 3?

> 3) drop the functionality related to the calling a callable to get the Param value at build time. I'm not a fan of this, as this would mean, at least for my use case, a lot more repeated code in some cases; users who want to execute a callable for the Param value at build time would be forced to pass it as a build kwarg every time.

adhtruong commented:
Sounds good! I'd say 3 if my preference with 2 if strong general use case depending what other @litestar-org/maintainers thinks
